src/scheduler/Executor.cpp

```diff
@@ -223,19 +237,33 @@ void Executor::threadPoolThread(int threadPoolIdx)
             releaseClaim();
         }
     }
-}
 
-void Executor::shutdownThreadPoolThread(int threadPoolIdx)
-{
-    const auto& logger = faabric::util::getLogger();
-    logger->debug("Shutting down thread pool thread {}:{}", id, threadPoolIdx);
-
-    threadPoolThreads.at(threadPoolIdx) = nullptr;
+    if (selfShutdown) {
+        logger->debug(
+          "Shutting down thread pool thread {}:{}", id, threadPoolIdx);
+
+        // Note - we have to keep a record of dead threads so we can join them
+        // all when the executor shuts down
+        std::shared_ptr<std::thread> thisThread =
+          threadPoolThreads.at(threadPoolIdx);
+        deadThreads.emplace_back(thisThread);
+
+        // Set this thread to nullptr
+        threadPoolThreads.at(threadPoolIdx) = nullptr;
+
+        // See if any threads are still running
+        bool isFinished = true;
+        for (auto t : threadPoolThreads) {
```

Comment:
I would love not to iterate through this vector checking equality, but all the stdlib options look pretty verbose, and this isn't on the critical path performance-wise so I'd rather stick with readability and debuggability).